### PR TITLE
pkp/pkp-lib#7649 Add support for allowed_hosts limits to avoid HOST header injection

### PR DESCRIPTION
**Pull request by asmecher**

### CODE REVIEW

**Comment by bozana** [comment]
@asmecher, where/how do we implicitly trust `base_url` hosts?

**Comment by bozana** [comment]
I would have two more questions, not closely connected with this issue -- they work, so you can ignore them:

1) In the function `getBaseUrl`: 
Why do we use `false` in `getServerHost` when the first parameter and the return value should be a string:
```
 $serverHost = $this->getServerHost(false);
        if ($serverHost !== false) {
        ...

```
2) In getServerHost:
Why don't we have `getServerHost($default = 'localhost'...` instead of:
```
getServerHost($default = null, $includePort = true)
    {
        if ($default === null) {
            $default = 'localhost';
        }
```

**Comment by asmecher** [comment]
> @asmecher, where/how do we implicitly trust base_url hosts?

What I mean is that when the `base_url[...]` configuration entries are being used, we use them in place of `getIndexUrl` (which in turn relies on the `HTTP_HOST` header) in order to construct URLs; see `PKPRouter::_urlGetBaseAndContext`. So using `base_url[...]` means we don't rely on `HTTP_HOST` when generating URLs.

> In the function `getBaseUrl`, why do we use false in `getServerHost` when the first parameter and the return value should be a string?

This is old code; it's checking whether a hostname could be automatically detected or not. By passing in a `false` as a default, then checking if that `false` is passed back out of the function, we're determining whether the default had to be used (i.e. nothing better could be identified for a hostname).

> In getServerHost, why don't we have `getServerHost($default = 'localhost'...` instead of...

This is also old code, but using
```php
function myFunction($param = null, $someOtherParameter = null) {
    if ($param === null) $param = 'someStringValue';
    ...
}
```
instead of
```php
function myFunction($param = 'someStringValue', $someOtherParameter = null) {
    ...
}
```
...can be useful to allow callers to avoid re-coding the `someStringValue` when they really just want to provide something for `$someOtherParameter` and leave `$param` at its default. This kind of hack can go away when we're allowed to use named parameters.

**Comment by bozana** [comment]
Thanks @asmecher, I would then say it can be merged :)

**Comment by jonasraoni** [comment]
@asmecher What about failing with a 400 + a helpful message when the host is invalid? Just to avoid someone pointing a fake domain to the installation.

Perhaps it would be useful to also allow the rule to be based on a regular expression. Then the user would be able to define the allowed domains using something like `/^[\w]+\.pkp\.sfu\.ca$/`.

**Comment by asmecher** [comment]
Thanks, @bozana and @jonasraoni!

> What about failing with a 400 + a helpful message when the host is invalid? Just to avoid someone pointing a fake domain to the installation.

This led me to a bit of a rewrite. The earlier implementation had operated when a URL was generated (e.g. when generating a response), which is a bad place to suddenly jump in and respond with a `400`. So instead I wrote a new authorization policy, which is a much more logical place to check and intervene. I've updated all the PRs to suit.

@bozana and @jonasraoni, could you take another look? Much appreciated!

File: classes/security/authorization/AllowedHostsPolicy.inc.php [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+            AuthorizationPolicy::AUTHORIZATION_PERMIT : AuthorizationPolicy::AUTHORIZATION_DENY;
+    }
+
+    /**
+     * Handle a mismatch in the allowed hosts expectation.
+     */
+    public function callOnDeny()
+    {
+        http_response_code(400);
+        fatalError('400 Bad Request');
```
Comment (jonasraoni) [inline classes/security/authorization/AllowedHostsPolicy.inc.php:70]: I think it's useful to log a helpful error message here, just to help the admin figure out why he's getting a 400.

**Review by jonasraoni — APPROVED** [review]

**Comment by bozana** [comment]
@asmecher, sorry for my 'silly' question 'again': What if the host is one of the base_url hosts now that the check is at the beginning? Would PKPHandler authorization work? :flushed:
Else, yes, you can merge... :-)

**Comment by asmecher** [comment]
> What if the host is one of the base_url hosts now that the check is at the beginning? Would PKPHandler authorization work?

@bozana, not a silly question at all! Now that this is implemented as a policy/request validation concern rather than a URL generation concern, it'll be necessary for either...
- all domains to be explicitly listed in `allowed_hosts`, or
- `allowed_hosts` checking to be disabled by leaving the setting empty. This can actually be done safely when virtual hosting, since the web server will already be using the `Host` header to map the request to the site, essentially filtering the header for allowed values (**if that's the way virtual hosting is configured**). In that scenario there would be no need to use `allowed_hosts`.

File: classes/security/authorization/AllowedHostsPolicy.inc.php (hunk 1 above)
Comment (asmecher) [reply classes/security/authorization/AllowedHostsPolicy.inc.php:70]: Good idea, thanks -- I will add this.

**Pull request merged**